pyro/infer/autoguide/guides.py

```diff
@@ -321,6 +350,13 @@ def __init__(self, model, prefix="auto", init_loc_fn=init_to_median):
         model = InitMessenger(init_loc_fn)(model)
         super().__init__(model, prefix=prefix)
 
+    def _init_params(self):
+        self.loc = self._init_loc()
+
+    @constraint
```

Comment:
We shouldn't need a constraint for loc, it can be unconstrained.

Reply:
This is to register them as `nn.Parameters`, otherwise these won't be captured by `pyro.module`. Alternatively, we can have do `nn.Parameter(self._init_loc())`?

Reply:
Changed to `nn.Parameter` - it seems a little odd to have some parameters as `nn.Parameter` and others as regular tensors which are converted into `nn.Parameter` by the constraint descriptor later, but it works.

Reply:
Hmm, good point. Maybe it's cleanest to always set `self.param = nn.Parameter(...)` even if we know that param is constrained.

Reply:
I tried that. 😄 The issue with that is that if we set `self.scale = nn.Parameter(..)`, it does not go through `ConstraintDescriptor.__set__` and the constraint isn't registered. My guess is that it gets intercepted by `__set__` from `nn.Module`, but I need to look into that.

Reply:
If we would like to preserve the same overriding behavior for parameters that are initialized as tensors or `nn.Parameter` alike, one option is to create an intermediate class like `AutoGuideModule` (likewise, for `ConstrainedAutoGuideModule`), otherwise any `nn.Parameter` attributes will be set by `nn.Module.__setattr__` without triggering the `__set__` method of the ConstraintDescriptor (and hence no constraints will be set).

```python
class AutoGuideModule(AutoGuide, nn.Module):
    def __setattr__(self, key, value):
        if key in type(self).__dict__ and inspect.isdatadescriptor(getattr(type(self), key)):
            getattr(type(self), key).__set__(self, value)
        else:
            super().__setattr__(key, value)
```

If we do this, we can probably also just have `forward = __call__` for this class.